File: compiler/src/queries/PlaceholderQueries.swift
extension Queries {
    public static var placeholder: Query<Node, [Node], ConstructedType?> {
        { db, node, body in
            guard db.contains(IsPlaceholder.self, for: node),
                let group = db[node, Typed.self]?.group
            else { return }

            let others = group.nodes.filter { $0 != node }
            let type = group.types.isEmpty ? nil : group.types[0]

            body(node, others, type)
        }
    }
}

File: compiler/queries/src/placeholder_queries.rs
use crate::QueryCtx;
use wipple_core::{
    db::Node,
    typecheck::{groups::Typed, ty::ConstructedTy},
};
use wipple_syntax::expressions::placeholder_expression::IsPlaceholder;

pub fn placeholder<'a>(
    db: &QueryCtx<'a>,
    node: Node,
) -> Option<(Node, Vec<Node>, Option<&'a ConstructedTy>)> {
    if !db.contains::<IsPlaceholder>(node) {
        return None;
    }

    let Typed(Some(group)) = db.get(node)? else {
        return None;
    };

    Some((
        node,
        group
            .nodes
            .iter()
            .copied()
            .filter(|other| *other != node)
            .collect(),
        group.tys.first(),
    ))
}
